xls/modules/dbe/lz4_decoder.x

```diff
@@ -0,0 +1,618 @@
+
+        // Reset, set or increment read pointer
+        let rd_ptr = if rx_reset {
+            uN[PTR_WIDTH]:0
+        } else if rx_cp {
+            state.wr_ptr - rx.cp_off - uN[PTR_WIDTH]:1
+        } else if is_copying {
+            state.rd_ptr + uN[PTR_WIDTH]:1
+        } else {
+            state.rd_ptr
+        };
+
+        // Check whether rd_ptr points to an unwritten HB cell
+        let rd_ptr_valid = (rd_ptr < state.wr_ptr) || state.hb_all_valid;
+
+        // HB READ - read the symbol from history buffer
+        let do_hb_read = emit_sym_from_hb && rd_ptr_valid;
+        let tok = send_if(tok, o_ram_hb_rd_req, do_hb_read,
+            ram::ReadWordReq<u32:1>(rd_ptr));
+        let (tok, hb_rd) = recv_if(tok, i_ram_hb_rd_resp, do_hb_read,
+            RamZeroReadResp<SYM_WIDTH>());
+
+        // Generate ERROR_INVAL_CP when we should've read from HB but
+        // we couldn't because CP points to an invalid location
+        let err_inval_cp = emit_sym_from_hb && !do_hb_read;
+
+        // Send decoded data symbol or Mark
+        let zero_data = dbe::zero_plain_data<SYM_WIDTH>();
+        let (data_valid, data) = if rx_reset {
+            // Propagate RESET in all states
+            (true, PlainData {
+                is_mark: true,
+                mark: Mark::RESET,
+                ..zero_data
+            })
+        } else if state.err {
+            // Do not send anything else while in an error state
+            (false, zero_data)
+        } else if rx_error || rx_end {
+            // Propagate ERROR and END tokens
+            (true, PlainData {
+                is_mark: true,
+                mark: rx.mark,
+                ..zero_data
+            })
+        } else if rx_unexpected_mark {
+            // Generate ERROR_BAD_MARK
+            (true, PlainData {
+                is_mark: true,
+                mark: Mark::ERROR_BAD_MARK,
+                ..zero_data
+            })
+        } else if err_inval_cp {
+            // Generate ERROR_INVAL_CP
+            (true, PlainData {
+                is_mark: true,
+                mark: Mark::ERROR_INVAL_CP,
+                ..zero_data
+            })
+        } else if emit_sym_lit {
+            // Replicate symbol from LT token
+            (true, PlainData {
+                is_mark: false,
+                data: rx.lt_sym,
+                ..zero_data
+            })
+        } else if emit_sym_from_hb {
+            // Replicate symbol from HB
+            (true, PlainData {
+                is_mark: false,
+                data: hb_rd.data,
+                ..zero_data
+            })
+        } else {
+            (false, zero_data)
+        };
+        let tok = send_if(tok, o_data, data_valid, data);
+
+        // HB WRITE - write emitted symbol to the history buffer
+        let do_hb_write = data_valid && !data.is_mark;
+        let tok = send_if(tok, o_ram_hb_wr_req, do_hb_write,
+            ram::WriteWordReq<u32:1>(state.wr_ptr, data.data));
```

Comment:
LZ4 is a byte-oriented compression right?  Should there be some use of masks to handle a WordReq'ed sized write?

Reply:
I think a TODO is warranted here then.  I think most RAMs we'll have available for implementation would have a data bus of greater than 8 bits.

Reply:
I can add a TODO. But do we want to make DSLX code aware of the port size of a specific RAM IP used in the hardware?  E.g. in FPGAs you'd frequently rely on RAM block being auto-inferred from the Verilog code, thus the synthesis tool takes care of a mismatch between the data/address bus widths of available hard-IP core and the expectations set by the RTL design.

Reply:
I think it should be an input parameter.  SRAMS take a non-trivial amount of time to generate collateral and incorporate during an ASIC desgin flow.